language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java

```diff
@@ -0,0 +1,523 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Generates Service Template Snippet completion items.
+ *
+ * @since 2.0.0
+ */
+public class ServiceTemplateGenerator {
+
+    private static final LanguageServerContext.Key<ServiceTemplateGenerator> SERVICE_TEMPLATE_GENERATOR_KEY =
+            new LanguageServerContext.Key<>();
+
+    private Map<Pair<String, String>, List<ListenerMetaData>> moduleListenerMetaDataMap;
+
+    private boolean isInitialized;
+
+    public boolean initialized() {
+        return isInitialized;
+    }
+
+    private ServiceTemplateGenerator(LanguageServerContext context) {
+        context.put(SERVICE_TEMPLATE_GENERATOR_KEY, this);
+        this.moduleListenerMetaDataMap = new ConcurrentHashMap<>();
+        CompletableFuture.runAsync(() -> initialize(context));
+    }
+
+    /**
+     * Get an instance of the Service Template Generator.
+     *
+     * @param context Language Server Context.
+     * @return {@link ServiceTemplateGenerator} Service Template Generator.
+     */
+    public static ServiceTemplateGenerator getInstance(LanguageServerContext context) {
+        ServiceTemplateGenerator serviceTemplateGenerator =
+                context.get(SERVICE_TEMPLATE_GENERATOR_KEY);
+        if (serviceTemplateGenerator == null) {
+            serviceTemplateGenerator = new ServiceTemplateGenerator(context);
+        }
+        return serviceTemplateGenerator;
+    }
+
+    /**
+     * Initializes the Service Template Generator.
+     *
+     * @param context Language Server Context.
+     */
+    private void initialize(LanguageServerContext context) {
+        if (!this.isInitialized) {
+            loadListenersFromDistribution(context);
+            this.isInitialized = true;
+        }
+    }
+
+    /**
+     * Given a module symbol, find and populate service metadata into the moduleServiceTemplateMap cache.
+     * Used to dynamically add new entries to the cache.
+     *
+     * @param moduleSymbol Module symbol.
+     * @param orgName      Organization name of the module symbol.
+     * @param moduleName   Module name of the module symbol.
+     * @param modulePrefix Module Prefix of the modules symbol.
+     * @param shouldImport Whether import text should be added.
+     * @param ctx          BallerinaCompletion context.
+     * @return {@link List<LSCompletionItem>} Set of completion items corresponding to the listeners
+     * in the given module.
+     */
+    public synchronized List<LSCompletionItem> generateAndPopulate(ModuleSymbol moduleSymbol,
+                                                                   String orgName,
+                                                                   String moduleName,
+                                                                   String modulePrefix,
```

Comment:
Can't we deduce the `orgName`, `moduleName` and `modulePrefix` from the `moduleSymbol` itself?

Reply:
Done